staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types_discovery.go

```diff
@@ -0,0 +1,132 @@
+limitations under the License.
+*/
+
+package v1
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// APIGroupDiscoveryList is a resource containing a list of APIGroupDiscoveries.
+// The shape is chosen to be very similar to "normal" api endpoints and work well with our type system
+// This is what is returned from the /discovery/v2 endpoint.
+type APIGroupDiscoveryList struct {
+	TypeMeta `json:",inline"`
+	// ResourceVersion will not be set, because this does not have a replayable ordering among multiple apiservers.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// Groups is the list of groups for discovery.
+	// +listType=map
+	// +listMapKey=name
+	Groups []APIGroupDiscovery `json:"groups" protobuf:"bytes,2,rep,name=groups"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// APIGroupDiscovery holds information about which resources are being served for all version of the API Group.
+type APIGroupDiscovery struct {
+	TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	// The only field completed will be Name.  For instance, resourceVersion will be empty.
+	// Name is the name of the API group whose discovery information is presented here.
+	// Name is allowed to be "" to represent the legacy, ungroupified resources.
+	// This is included to make the standard deserialization easier.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// Versions are the versions supported in this group. They are sorted in descending order of preference,
+	// with the preferred version being the first entry.
+	// +listType=map
+	// +listMapKey=version
+	Versions []APIVersionDiscovery `json:"versions,omitempty" protobuf:"bytes,2,rep,name=versions"`
+}
+
+// APIVersionDiscovery holds a list of APIResourceDiscovery types that are served for a particular version within an API Group.
+type APIVersionDiscovery struct {
+	// Version is the name of the version within a group version.
+	Version string `json:"version" protobuf:"bytes,1,opt,name=version"`
+	// Resources is a list of APIResourceDiscovery objects for the corresponding group version.
+	// +listType=map
+	// +listMapKey=resource
+	Resources []APIResourceDiscovery `json:"resources,omitempty" protobuf:"bytes,2,rep,name=resources"`
+	// DocumentStatus marks whether a group version's discovery document is up to date.
+	// "Current" indicates no problems when fetching the discovery document. "Stale" indicates
+	// that there was an error fetching the discovery document, and the current version may not
+	// be up to date.
+	DocumentStatus DocumentStatus `json:"documentStatus,omitempty" protobuf:"bytes,3,opt,name=documentStatus"`
```

Comment:
What do you think about the name "Freshness"?

Reply:
renamed to  `DiscoveryFreshness`

Reply:
Why the `Discovery` prefix?

Reply:
In case metav1 has other `Freshness` related enums in the future. Would it be better to have something like 
`Freshness DiscoveryFreshness json:...` instead of `DiscoveryFreshness DiscoveryFreshness json:...`?

Reply:
Yeah, I like it more, "Discovery" in the var name doesn't really add much data here :-)

Reply:
Updated to `Freshness DiscoveryFreshness`